pkg/query/endpointset_test.go

```diff
@@ -345,12 +385,66 @@ func TestEndpointSetUpdate(t *testing.T) {
 			defer endpoints.Close()
 
 			discoveredEndpointAddr := endpoints.EndpointAddresses()
-			endpointSet := makeEndpointSet(discoveredEndpointAddr, tc.strict, time.Now)
+			var endpointSet *EndpointSet
+			// specify only "store_type" to exclude "external_labels"
+			if tc.name == "no external labels" {
+				endpointSet = makeEndpointSet(discoveredEndpointAddr, tc.strict, time.Now, "store_type")
+			} else {
+				endpointSet = makeEndpointSet(discoveredEndpointAddr, tc.strict, time.Now)
+			}
 			defer endpointSet.Close()
 
 			endpointSet.Update(context.Background())
 			testutil.Equals(t, tc.expectedEndpoints, len(endpointSet.GetEndpointStatus()))
 			testutil.Equals(t, tc.expectedEndpoints, len(endpointSet.GetStoreClients()))
+			// slow or unavailable endpoint should collect nothing
+			if tc.name == "slow endpoint" || tc.name == "unavailable endpoint" {
+				testutil.Ok(t, promtestutil.CollectAndCompare(endpointSet.endpointsMetric, strings.NewReader("")))
+				return
+			}
+
+			if tc.name == "no external labels" {
+				expectedMetrics := fmt.Sprintf(
+					`
+					# HELP thanos_store_nodes_grpc_connections Number of gRPC connection to Store APIs. Opened connection means healthy store APIs available for Querier.
+					# TYPE thanos_store_nodes_grpc_connections gauge
+					thanos_store_nodes_grpc_connections{store_type="sidecar"} %d
+					`,
+					tc.expectedEndpoints,
+				)
+				testutil.Ok(t, promtestutil.CollectAndCompare(endpointSet.endpointsMetric, strings.NewReader(expectedMetrics)))
+				return
+			}
+
+			var externalLabels string
+			if tc.name == "long external labels" {
+				externalLabels = strings.Repeat(`lbl="val", `, 1000)
+				externalLabels = externalLabels[:len(externalLabels)-2]
+			} else {
+				externalLabels = fmt.Sprintf(`a="b", addr=%q`, discoveredEndpointAddr[0])
+			}
+			// labels too long must be trimmed
```

Comment:
Please fix all comments (:

Reply:
ok. just to be clear, you mean I should add more comments?

Reply:
oh. make them sentences. I got it. Fixed.